### ISSUE TITLE
Some dtype values are not supported for some transformations

### ISSUE THREAD

**Issue by dipspb** [aleju/imgaug#67]
Set of unsupported dtype values is depend on a list of transformation used in augmentation.

Here is result of test:
> ['Linux 092c6277fddb 4.4.0-93-generic #116-Ubuntu SMP Fri Aug 11 21:17:51 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux']
> /root/miniconda3/envs/ml-toolset/bin/python
> 3.5.2 | packaged by conda-forge | (default, Jan 19 2017, 15:28:33) 
> [GCC 4.8.2 20140120 (Red Hat 4.8.2-15)]
> NumPy: 1.12.1
> imgaug: 0.2.4
> 
> TYPE: numpy.int8 PASS: (32, 64, 64, 3)
> TYPE: numpy.int16 PASS: (32, 64, 64, 3)
> TYPE: numpy.int32 PASS: (32, 64, 64, 3)
> TYPE: numpy.int64 PASS: (32, 64, 64, 3)
> TYPE: numpy.uint8 PASS: (32, 64, 64, 3)
> TYPE: numpy.uint16 PASS: (32, 64, 64, 3)
> TYPE: numpy.uint32 EXCEPTION: src data type = 6 is not supported
> TYPE: numpy.uint64 PASS: (32, 64, 64, 3)
> TYPE: numpy.float16 EXCEPTION: src data type = 23 is not supported
> TYPE: numpy.float32 PASS: (32, 64, 64, 3)
> TYPE: numpy.float64 PASS: (32, 64, 64, 3)
> TYPE: numpy.float128 EXCEPTION: array type 13 not supported
> 
Here is a source code used:
```
import sys
import numpy as np

import imgaug as ia
from imgaug import augmenters as iaa

ia.seed(1)

seq = iaa.Sequential([
    iaa.Fliplr(0.5), # horizontal flips
    iaa.Crop(percent=(0, 0.1)), # random crops
    iaa.Sometimes(0.5,
        iaa.GaussianBlur(sigma=(0, 0.5))
    ),
    iaa.ContrastNormalization((0.75, 1.5)),
    iaa.AdditiveGaussianNoise(loc=0, scale=(0.0, 0.05*255), per_channel=0.5),
    iaa.Multiply((0.8, 1.2), per_channel=0.2),
    iaa.Affine(
        scale={"x": (0.8, 1.2), "y": (0.8, 1.2)},
        translate_percent={"x": (-0.2, 0.2), "y": (-0.2, 0.2)},
        rotate=(-25, 25),
        shear=(-8, 8)
    )
], random_order=True) # apply augmenters in random order

\# Here is [:-1] and here is no sctypes['uint']
\# because sctypes['int'] on python 3.5 NumPy-1.12.1 has
\# int list extended with uint list and it has uint list appended
types_to_test = np.sctypes['int'][:-1] + np.sctypes['float']

def type_name(t):
    return str(t).split("'")[1]

uname = !uname -a
print(uname)
print(sys.executable)
print(sys.version)
print('NumPy: ' + np.__version__)
print('imgaug: ' + ia.__version__)
print()

for t in types_to_test:
    t_name = type_name(t)
    images = np.array(np.random.rand(32, 64, 64, 3), dtype = t)
    try:
        images_aug = seq.augment_images(images)
        print('TYPE: {} PASS: {}'.format(t_name, images_aug.shape))
    except:
        e = sys.exc_info()[1]
        print('TYPE: {} EXCEPTION: {}'.format(t_name, e))
```

**Comment by aleju** [comment]
Hey, thanks for the test script. I'll have to add that to the other unit tests.
Currently the library is primarily focused on `uint8` dtypes. Others may or may not work. Usually it is best to convert everything to uint8. Note that even if an augmenter works with other dtypes without an error, they usually all expect the values to be in the range 0-255. For e.g. ContrastNormalization other value ranges would be quite problematic as it expects the mean pixel intensity to be 128 and measures the difference from that value.

What's your use case for using other dtypes? There will likely be extra functions at some point for segmentation maps and heatmaps.

**Comment by amirbar** [comment]
Not sure what exactly is your use case, but if it is mainly rotations, cropping and translations - the fork below is working good with any dtype. 

https://github.com/amirbar/imgaug

@aleju do you think we can just preserve the dtype of the input images and raise exception on op's that do not support dtype != uint8?
are you interested in such PR?

**Comment by aleju** [comment]
Yes, it would be better to preserve dtypes in the current state of the library.
Btw, I looked at your fork and you have used `dtype=images.dtype` a few times. But `images` (and its dtype) can change during the augmentation process, as `_augment_images()` works in-place. So I think (without testing it) you should save the dtype via e.g. `input_dtype = images.dtype` somewhere at the start of the function call and then use `dtype=input_dtype` later.

**Comment by amirbar** [comment]
Thanks, will do it 👍

**Comment by dipspb** [comment]
Hi @aleju ,

> What's your use case for using other dtypes?

When I missed dtype parameter while creating ndarray, next moment I suddenly have issue found. Besides missed dtype also seems it had a wrong shape (I attempted to parse image ndarray from pandas DataFrame).

I got some error message that was totally misleading. It was something like "blah-blah-blah interpolation is zero" from inside of cv2. I attempted to reproduce it w/o any usage of pandas and have no success with it. Instead I discovered it behaves in a different way for different dtype values.

Then I decided to report it here as considering it could be useful to implement it in a some common way for all dtype values or limit them just to uint8.

**Comment by amirbar** [comment]
@aleju would you like me to come with a PR considering your comment above?

**Comment by aleju** [comment]
Yeah, feel free to add a PR.

**Comment by tartavull** [comment]
+1
If you are doing a semantic segmentation and have more than 256 classes it's natural to have uint16 images that you want to apply the same transformations.

**Comment by Shreeyak** [comment]
+1
I'm working with surface normals - converting them into an uint16 RGB image for augmentations.

In the current release, I do see in the documentation that int16 is supported, as well as other data types for most operations: https://imgaug.readthedocs.io/en/latest/source/api_augmenters_blur.html

**Comment by aleju** [comment]
The dtype support has been improved quite a lot since this thread was started.
The page https://imgaug.readthedocs.io/en/latest/source/dtype_support.html gives an overview of supported dtypes. uint8 is still the most reliable and unittested one, but other dtypes should often work too and there is at least rudimentary unittesting for them.

**Comment by Shreeyak** [comment]
Wow, that's amazing!
Thanks! I'm sure this will be useful info for others who come across this issue.
